vm/devices/storage/disk_nvme/nvme_driver/src/queue_pair.rs

```diff
@@ -280,43 +303,100 @@ impl Issuer {
             .probe_gpns(mem.gpns())
             .map_err(RequestError::Memory)?;
 
-        let prp = if mem
-            .gpns()
-            .iter()
-            .all(|&gpn| guest_memory.iova(gpn * PAGE_SIZE64).is_some())
+        // TODO: add check if the memeory if VA-backed.
+        let is_va_backed = false;
+        let (prp, is_memory_pinned) = if !is_va_backed
+            && mem
+                .gpns()
+                .iter()
+                .all(|&gpn| guest_memory.iova(gpn * PAGE_SIZE64).is_some())
         {
             // Guest memory is available to the device, so issue the IO directly.
-            self.make_prp(
-                mem.offset() as u64,
-                mem.gpns()
-                    .iter()
-                    .map(|&gpn| guest_memory.iova(gpn * PAGE_SIZE64).unwrap()),
+            (
+                self.make_prp(
+                    mem.offset() as u64,
+                    mem.gpns()
+                        .iter()
+                        .map(|&gpn| guest_memory.iova(gpn * PAGE_SIZE64).unwrap()),
+                )
+                .await,
+                false,
             )
-            .await
         } else {
-            tracing::debug!(opcode = opcode.0, size = mem.len(), "double buffering");
-
-            // Guest memory is not accessible by the device. Double buffer
-            // through an allocation.
-            let double_buffer_pages = double_buffer_pages.insert(
-                self.alloc
-                    .alloc_bytes(mem.len())
-                    .await
-                    .ok_or(RequestError::TooLarge)?,
-            );
+            // Guest memory is not accessible by the device.
+            // If guest memory is VA-backed, and IO size exceeds threshold, pin the memory.
+            let mut prp_result = None;
+            let mut is_pinned = false;
+            if let Some(io_threshold) = self.io_threshold {
+                if is_va_backed && self.partition.is_some() && mem.len() as u32 > io_threshold {
```

Comment:
I would rather keep is_va_backed out of this routine. The pinning code (whatever form that takes) should take care of this.